libs/contrib/Control/Validation.idr

```diff
@@ -0,0 +1,193 @@
+    left <|> right = MkValidator \x => MkEitherT $ do
+        case !(runEitherT $ validateT left x) of
+            (Right r) => pure $ Right r
+            (Left e) => case !(runEitherT $ validateT right x) of
+                (Right r) => pure $ Right r
+                (Left e') => pure $ Left (e <+> " / " <+> e')
+
+||| Alter the input before validation using given function.
+export
+contramap : (a -> b) -> ValidatorT m b c -> ValidatorT m a c
+contramap f v = MkValidator (validateT v . f)
+
+
+||| Given a value x and a decision procedure for property p, validateT if p x
+||| holds, returning a proof if it does. The procedure also has access to the
+||| raw input in case it was helpful.
+export
+decide : Monad m => {0 a, b : Type} -> String -> (x : b) -> {p : b -> Type} -> (a -> (x : b) -> Dec (p x)) -> ValidatorT m a (p x)
+decide {a} {b} msg x {p} f = MkValidator $ \a => case f a x of
+    Yes prf => pure prf
+    No _ => fail msg
+
+||| Given a function converting a into Maybe b, build a Validator of a
+||| converting it into b.
+export
+fromMaybe : Monad m => (a -> String) -> (a -> Maybe b) -> ValidatorT m a b
+fromMaybe e f = MkValidator \a => case f a of
+    Nothing => fail $ e a
+    Just b => pure b
+
+||| Verify whether a String represents a natural number.
+export
+natural : Monad m => ValidatorT m String Nat
+natural = fromMaybe mkError parsePositive
+    where
+    mkError : String -> String
+    mkError s = "'" <+> s <+> "' is not a natural number."
+
+||| Verify whether a String represents an Integer
+export
+integral : (Num a, Neg a, Monad m) => ValidatorT m String a
+integral = fromMaybe mkError parseInteger
+    where
+    mkError : String -> String
+    mkError s = "'" <+> s <+> "' is not an integer."
+
+||| Verify that a string represents a decimal fraction.
+export
+double : Monad m => ValidatorT m String Double
+double = fromMaybe mkError parseDouble
+    where
+    mkError : String -> String
+    mkError s = "'" <+> s <+> "is not a decimal."
+
+
+||| Verify whether a list has a desired length.
+export
+length : Monad m => (l : Nat) -> ValidatorT m (List a) (Vect l a)
+length l = MkValidator (validateVector l)
+    where
+    validateVector : (l : Nat) -> List a -> Result m (Vect l a)
+    validateVector Z [] = pure []
+    validateVector (S _) [] = fail "Missing list element."
+    validateVector Z (_ :: _) = fail "Excessive list element."
+    validateVector (S k) (x :: xs) = do
+        ys <- validateVector k xs
+        pure (x :: ys)
+
+||| Verify that certain values are equal.
+export
+equal : (DecEq a, Monad m) => (x, y : a) -> ValidatorT m z (x = y)
+equal x y = MkValidator dec
+    where
+    dec : z -> Result m (x = y)
+    dec _ = case decEq x y of
+        Yes prf => pure prf
+        No _ => fail "Values are not equal."
+
+||| Verify that a Nat is less than or equal to  certain bound.
+export
+lteNat : Monad m => {0 a : Type} -> (bound, n : Nat) -> ValidatorT m a (LTE n bound)
```

Comment:
These look suspicious to me: the conclusion mentions an `a` that has absolutely
nothing to do with the validation at hand. And the `a` argument is indeed ignored
in `\_, x => isLTE x bound`.

A solution may be to have

```idris
data ValidatorT : (Type -> Type) -> (a : Type) -> (a -> Type) -> Type where
    MkValidator : ((x : a) -> Result m (p x)) -> ValidatorT m a p
```

and then we can have

```idris
lteNat : Monad m => (bound : Nat) -> ValidatorT m Nat (\ n => LTE n bound)
```

The simpler things like `natural` can be:

```idris
natural : Monad m => ValidatorT m String (\ _ => Nat)
```

Of course the problem is that it's not a functor / applicative / monad
anymore but I wonder whether we can recover useful combinators nonetheless.
You can still have `map` for instance:

```idris
map : (forall x. p x -> q x) -> ValidatorT m a p -> ValidatorT m a q
```

Reply:
You're right and I thought about it as well. But when I realised that it's no longer a monad that way, I gave up. I think `do` notation is too valuable here to resign on it easily. And if it's not a monad, it can't be a monad transformer, so you won't be able to mix validation with an IO operation (for instance validating a large input file on the fly). Besides, I thought, maybe it's not so bad? It might happen from time to time, that having the original `a` for validating some of it's parts might be useful. Although admittedly I couldn't come up with a convincing example so far.

All in all I'm not really sure what should be done with it. What you point out is troubling indeed, but the ability to combine validators easily is essential here (it's precisely why I wrote this module in the first place). Without `ValidatorT` being a monad I think most of the value of this module would be lost.

Reply:
It's in `contrib` so we can easily revisit this if we find a better design in
the future. In the meantime, I'd rather replace these `a` with `()`. It should
help with type inference & clarifies that we do not care about that `a`.

Reply:
This would be nice, but I believe there is still the same problem as before, because having validators `ValidatorT m a b` and `ValidatorT m () (p x)`, you can't bind them together anymore. To maintain composability, we would have to have a wrapper, that would convert `Validator m () (p x)` to a suitable type, so that it could be composed. This is possible of course, but seems somewhat superfluous…

Reply:
Oh I see. Looking at `ValidatorT` as a monad gives the `a` a `Reader`-flavour
rather than simply an input to the validator. That's... strange.

Maybe the right interface for composing `ValidatorT m` is more [`Arrow`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Arrow.html) than `Monad`!
We are, after all, working in the Kleisli cat of the `ResultT m` monad.

Reply:
Hmm, you're right. Arrow might be exactly what we want here. Making `ValdatorT` synonym of `Result` is not that great, because it makes you pass the input manually to all validators one after another. I consider this `Reader`-flavour quite handy in this context. Also I have never worked with arrows before, so it's a great opportunity to learn more. I'll read a little bit and try to remake `ValidatorT` as an arrow, because from the first lines [here](https://en.wikibooks.org/wiki/Haskell/Understanding_arrows) it already looks like it's exactly what we want. I suppose I'll come back in a few days with the solution remade.

Reply:
So I have started porting `Control.Category` and `Control.Arrow` from Idris to Idris 2. However, I have a strange problem with interface resolution in the first of them:

```
module Control.Category

import Data.Morphisms


public export
interface Category (cat : k -> k -> Type) where
  id  : cat a a
  (.) : cat b c -> cat a b -> cat a c

public export
Category Morphism where
  id                = Mor id
  -- disambiguation needed below, because unification can now get further
  -- here with Category.(.) and it's only interface resolution that fails!
  (Mor f) . (Mor g) = Mor $ Basics.(.) f g

public export
Monad m => Category (Kleislimorphism m) where
  id                        = Kleisli (pure . id)
  (Kleisli f) . (Kleisli g) = Kleisli $ \a => g a >>= f

infixr 1 >>>

(>>>) : Category cat => cat a b -> cat b c -> cat a c
(>>>) {cat} f g = g . f
```
This fails to compile because:
```
While processing right hand side of >>> at Control/Category.idr:26:1--27:1:
Can't find an implementation for Category ?cat at:
26	(>>>) {cat} f g = g . f
	                  ^^^^^
```
It would seem that Idris either does not see that `?cat` is of type `cat` or ignores the requirement that `cat` should implement `Category` interface. Also I'm not sure, why instances of  `Category` have type `k -> k -> Type`. If I change it to `Type -> Type -> Type`, then it works.